### ISSUE TITLE
Need more information about how git lfs migrate export should work.

### ISSUE THREAD

**Issue by TomaszSzt** [git-lfs/git-lfs#5427]
Hi,

**Short description**

I decided to move out off the git-lfs after over-using it for some time in not very wisely chosen scenarios. I did follow the manual about:

`git lfs migrate export --everything --include=*`

but have serious doubts if it works the way I need and I would like to receive some clarifications from You, dear developers.

**Test conditions**
`git lfs version`
`git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3; git 77deabdf)`
`git version 2.39.0.windows.2`
`git lfs env`
`git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3; git 77deabdf)`
`git version 2.39.0.windows.2`
`Endpoint=http://xxxxxxx/r/lfs_migrate_test.git/info/lfs (auth=basic)`
`LocalWorkingDir=D:\x\lfs`
`LocalGitDir=D:\x\lfs\.git`
`LocalGitStorageDir=D:\x\lfs\.git`
`LocalMediaDir=D:\x\lfs\.git\lfs\objects`
`LocalReferenceDirs=`
`TempDir=D:\x\lfs\.git\lfs\tmp`
`ConcurrentTransfers=8`
`TusTransfers=false`
`BasicTransfersOnly=false`
`SkipDownloadErrors=false`
`FetchRecentAlways=false`
`FetchRecentRefsDays=7`
`FetchRecentCommitsDays=0`
`FetchRecentRefsIncludeRemotes=true`
`PruneOffsetDays=3`
`PruneVerifyRemoteAlways=false`
`PruneRemoteName=origin`
`LfsStorageDir=D:\x\lfs\.git\lfs`
`AccessDownload=basic`
`AccessUpload=basic`
`DownloadTransfers=basic,lfs-standalone-file,ssh`
`UploadTransfers=basic,lfs-standalone-file,ssh`
`GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core`
`GIT_LFS_PATH=C:\Program Files\Git LFS`
`git config filter.lfs.process = "git-lfs filter-process"`
`git config filter.lfs.smudge = "git-lfs smudge -- %f"`
`git config filter.lfs.clean = "git-lfs clean -- %f"`

remote: `Git-blit 1.9.2` with some customization regarding logging.
running commands from `cmd.exe` command line

**Test scenario 1**
1.Create an empty repository, make `git lfs` to track some files:
`git lfs track *.dat`
2.Commit, push.
3.Inspect remote for LFS use indicators (present) and file-store for a presence of LFS file (present)
4.Run
`git lfs migrate export --everything --include=*`
`git lfs untrack *.dat`
`git push --force`
In my case push --force just hung after displaying all messages (seems to be semi random fluke), so I aborted it, tried git push then force again. Worked.
5.Inspect remote for LFS use indicators (correct, no longer present, history was overwritten) and file-store for a presence of LFS file (present, *incorrect*)

***Questions***
1.Should the file remain in file store? 
2.Is it up to server to drop them automatically or up to LFS to tell server to drop them. If Yes, how? 

My expectation was, that if I push-force to not use LFS anymore, file storage space should be released.
Observations:
- If I delete the repository on a server the storage space is released;
- `git lfs migrate info --everything` output is not affected by `migrate` command until I do `push --force` it.

**Test scenario 2**
1.Do alike but create a bit more complex, non linear branch tree. Push everything on remote (push --all) and inspect if branches do exists there and LFS is in use.
2.Checkout one of branches.
3.Run 
`git lfs migrate export --everything --include=*`
5.Observe, that all history was re-written (good), but remote branch tracking was lost. If You attempt to push a branch You have to do the --set-upstream again what may be confusing in case of complex remotes and complex non 1:1 mapping.
4.Push it onto remote with
`git push --all --force`
In my case it again have frozen, again had to Ctrl-C it and push it again. 
5.Inspect, that use of LFS is no longer indicated by a remote, but file-store is still in use.

***Questions***
1.Is the breaking of local-to-remote branch mapping is an expected behavior or am I doing something wrong?

**Summary**
My primary concern is: how to re-claim the file-store space of files which are `migrate export`'ed.

Thanks for good work.

**Comment by bk2204** [comment]
Hey,

The Git LFS protocol doesn't provide a way to delete objects, so there's no way that the LFS objects on the server can be deleted by the client.  It's possible for the server to do that automatically, but since it isn't very efficient to do so, I'm not aware of any servers that do that.  You'd need to find a server-specific method of doing that, and I don't know what Gitblit offers there.

In general, you should avoid doing the migration on a repository with working trees and instead do it on a bare repository cloned with `--mirror`.  Then you don't have remote-tracking branches, since you don't typically want to push those to the remote.

The hanging is unexpected, and it could be in Git, or the remote side, and we wouldn't be able to know without seeing a trace with `GIT_TRACE=1 GIT_TRANSPORT_TRACE=1 GIT_CURL_VERBOSE=1`.  In any event, since Git LFS is invoked _before_ the rest of the push, I would suggest that it's probably not Git LFS and is instead your Git server, which may not gracefully handle many refs being pushed at once.

**Comment by TomaszSzt** [comment]
Thanks for prompt response.

The `push --force` hang indeed seems to be in Gitblit issue since it barfs in logs with some NullPointer exceptions. 

If I understand Your response correctly, the best way of performing the "go away from LFS as it had never existed before" it is to mirror the original repository from a server, do the housekeeping locally, __remove__ the repository from server the server specific way, and then push the cleaned up mirror back?

If yes, this is a surprise of all surprises for me, since even a simple deleting a branch from a remote with some LFS objects inside of it will require that kind of operation. 

Darn, I think I did heavily misunderstood how the LFS acts together with GIT. I was thinking that it is far more transparent and that I can switch back and forth LFS/non LFS on local and remote without affecting any cloned repositories. My bad.

Thanks again for a response.

**Comment by bk2204** [comment]
No, that's not the case.  What I suggested is that you do `git clone --mirror https://HOSTNAME/REPO.git REPO`, modify `REPO` locally with `git lfs migrate export`, and then do a `git push --all --force`.  I would recommend that you not do the migration operation on a repository with a working tree.  You certainly _can_, but it's trickier to handle and it's easier to mess up unless you know what you're doing.  Considering the fact that you're having trouble with remote-tracking branches (which should not be pushed and therefore require you not to use `--all`), I'm suggesting that a mirror clone will be easier for an export.

You don't need to delete the remote repository at all.  However, to delete the LFS objects on the server, you need to use a server-specific method, since Git LFS doesn't offer one.
